nomad/heartbeat.go

```diff
@@ -172,21 +174,44 @@ func (h *nodeHeartbeater) invalidateHeartbeat(id string) {
 	}
 }
 
-func (h *nodeHeartbeater) shouldDisconnect(id string) bool {
+func (h *nodeHeartbeater) disconnectState(id string) (bool, bool) {
+	node, err := h.State().NodeByID(nil, id)
+	if err != nil {
+		h.logger.Error("error retrieving node by id", "error", err)
+		return false, false
+	}
+
+	// Exit if this is the node already in a state other than ready.
+	if node.Status != structs.NodeStatusReady {
+		return false, false
+	}
+
 	allocs, err := h.State().AllocsByNode(nil, id)
 	if err != nil {
 		h.logger.Error("error retrieving allocs by node", "error", err)
-		return false
+		return false, false
 	}
 
 	now := time.Now().UTC()
+	// Check if the node has any allocs that are configured with max_client_disconnect,
+	// and if so, whether it has pending reconnects. If one alloc isn't yet expired
+	// it still has pending reconnects.
+	canDisconnect := false
+	hasPendingReconnects := false
 	for _, alloc := range allocs {
-		if alloc.DisconnectTimeout(now).After(now) {
-			return true
+		if !canDisconnect && alloc.DisconnectTimeout(now).After(now) {
+			canDisconnect = true
+		}
+		if !hasPendingReconnects && alloc.DesiredStatus == structs.AllocDesiredStatusRun && !alloc.Expired(now) {
+			hasPendingReconnects = true
+			break
+		}
```

Comment:
Is an alloc that meets the second case always guaranteed to be true for the first case? If not, breaking early creates an ordering problem where the order of allocs we iterate matters. If so, maybe we can make that relationship explicit here by moving this conditional under the first one? We know we can only hit the `hasPendingReconnects` once, so we can reduce it down to something like:

```go
var canDisconnect bool
for _, alloc := range allocs {
	if !canDisconnect && alloc.DisconnectTimeout(now).After(now) {
		canDisconnect = true
		if alloc.DesiredStatus == structs.AllocDesiredStatusRun && !alloc.Expired(now) {
			return true, true
		}
	}
}
return canDisconnect, false
```

Reply:
We can't do exactly this. You can only have pending reconnects if you are configured with a disconnect. So yes, if something evaluates to true for the second case, then by definition it needs to evaluate true for the first. However, you've made me realize I'm conflating nodes with allocs here. I'll rework this and resubmit.